src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandBinder.cs

```diff
@@ -152,40 +153,116 @@ internal Expression CreatePropertyNameExpression(IEdmEntityType elementType, IEd
             return Expression.Constant(property.Name);
         }
 
-        internal Expression CreatePropertyValueExpression(IEdmEntityType elementType, IEdmProperty property, Expression source)
+        internal Expression CreatePropertyValueExpression(IEdmStructuredType elementType, IEdmProperty property, Expression source)
         {
-            Contract.Assert(elementType != null);
             Contract.Assert(property != null);
             Contract.Assert(source != null);
 
-            return CreatePropertyValueExpressionWithFilter(elementType, property, source, filterClause: null);
+            return CreatePropertyValueExpressionWithFilter(elementType, property, source, null);
         }
 
-        internal Expression CreatePropertyValueExpressionWithFilter(IEdmEntityType elementType, IEdmProperty property,
-            Expression source, FilterClause filterClause)
+        internal Expression CreatePropertyValueExpressionWithFilter(IEdmStructuredType elementType, IEdmProperty property,
+            Expression source, ExpandedReferenceSelectItem expandItem)
         {
             Contract.Assert(elementType != null);
             Contract.Assert(property != null);
             Contract.Assert(source != null);
 
-            IEdmEntityType declaringType = property.DeclaringType as IEdmEntityType;
-            Contract.Assert(declaringType != null, "only entity types are projected.");
+            FilterClause filterClause = expandItem != null ? expandItem.FilterOption : null;
+
+            IEdmStructuredType declaringType = property.DeclaringType;
+
+            //if there is a path in $expand then find the right declaring type;
+            if (expandItem != null)
+            {
+                foreach (ODataPathSegment segment in expandItem.PathToNavigationProperty)
+                {
+                    PropertySegment propertyAccessPathSegment = segment as PropertySegment;
+                    if (propertyAccessPathSegment != null)
+                    {
+                        IEdmProperty newProperty = propertyAccessPathSegment.Property;
+                        declaringType = newProperty.DeclaringType;
+                        break;
```

Comment:
>break [](start = 24, length = 5)

You'll have to explain this logic to me.  If my path has multiple complex types, won't this stop at the first one? Is that what I want?